### DISCUSSION TITLE
TensorFlow Docker requirements

### DISCUSSION THREAD

**Discussion by vukitoso** [NVIDIA/open-gpu-kernel-modules#250] [Q&A] [answered]
Hello.
I plan to run TensorFlow programs in Docker containers.
The instructions at https://www.tensorflow.org/install/docker say:

`
Docker is the easiest way to enable TensorFlow GPU support on Linux since only the NVIDIA® GPU driver is required on the host machine (the NVIDIA® CUDA® Toolkit does not need to be installed).
`

Is "NVIDIA Linux Open GPU Kernel Module Source" (https://github.com/NVIDIA/open-gpu-kernel-modules) a full replacement for proprietary *.run files (e.g. NVIDIA-Linux-x86_64-510.73.05.run )?

**Comment by YusufKhan-gamedev** [comment]
AFAIK you need the CUDA toolkit on the docker container but not on the host system.

**Reply by vukitoso** [reply]
And what to install on the host: a proprietary version of * .run or can it be built from source?

**Reply by YusufKhan-gamedev** [reply]
I dont believe it matters AFAIK... but you can just build this from source so id go with that if I were you.

**Comment by PAR2020** [comment]
If you follow the README.md file, it does walk through how to build.
Regarding dockers and TensorFlow, if the closed source driver 515.43.04 driver works, then the open source should also. There are dependencies on CUDA for this driver version that need to be maintained across the host and docker. I'm checking to see if a 515.43.04 docker with compatible CUDA 11.7 Toolkit has been released.
For TensorFlow and docker related questions, check here:
https://github.com/NVIDIA/tensorflow
and
https://github.com/NVIDIA/nvidia-docker

**Comment by elezar** [comment, marked as answer]
In order to run GPU-enabled docker containers the [NVIDIA Container Toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/overview.html) is required. This ensures that the required driver libraries and devices are injected into the container as it is started. It is agnostic of the container runtime used (although it may required specific configuration for particular runtimes) and also allows the CUDA Toolkit to be bundled in the container image and be used as long as it is supported by the driver version on the host. In general, older CUDA versions are supported by newer driver versions and there is some forward-compatibility at least across minor versions that is also supported.

A CUDA-capable NVIDIA GPU driver is a prerequisite for the NVIDIA Container Toolkit, but it does not prescribe how the driver is installed. With this in mind, you should be able to select the installation mechanism that suits your use case.

I don't know whether this answers your original question of:

> Is "NVIDIA Linux Open GPU Kernel Module Source" (https://github.com/NVIDIA/open-gpu-kernel-modules) a full replacement for proprietary *.run files (e.g. NVIDIA-Linux-x86_64-510.73.05.run )?

But I would assume that the answer is generally "no", since only the kernel modules are open sourced, and from the perspective of the NVIDIA Container Toolkit at least, the "NVIDIA GPU driver" includes the NVIDIA Management Library (`libnvidia-ml.so`) which is used to enumerate devices for injection into the container.

**Reply by vukitoso** [reply]
thx

**Reply by PAR2020** [reply]
Thanks Ethan!
